src/Containers.jsx

```diff
@@ -40,19 +41,126 @@ class Containers extends React.Component {
         }));
     }
 
-    //TODO
-    stopContainer(container) {
-        return undefined;
+    stopContainer(container, timeout) {
+        document.body.classList.add('busy-cursor');
+        const id = container.ID;
+        if (!timeout) {
+            timeout = 10;
+        }
+        utils.varlinkCall(utils.PODMAN, "io.podman.StopContainer", JSON.parse('{"name":"' + id + '","timeout":' + timeout + '}' ))
+            .then(reply => {
+                const idStop = reply.container;
+                //update container info after stop
+                utils.varlinkCall(utils.PODMAN, "io.podman.InspectContainer", JSON.parse('{"name":"' + idStop + '"}'))
+                    .then(reply => {
+                        const newElm = JSON.parse(reply.container)
+                        //replace list with updated info
+                        let oldContainers = this.props.containers;
+                        let idx = oldContainers.findIndex((obj => obj.ID == idStop));
+                        oldContainers[idx] = newElm;
```

Comment:
This is invalid. `oldContainers` is the same object as `this.props.containers` and you must not update your own properties. Also, this is generally a bad and brittle architecture: Don't *assume* what the method you are calling (`StopContainer()` are doing - they might fail on any number of things. Just call `StopContainer` and then get the updated state for it with the already existing code.

If `class Containers` is actually supposed to manage and change the "containers" list, then this needs to be state, not a property. A property is owned and managed by the containing element, and then you just present it without ever touching it.

Reply:
Sorry, a little bit confusing. this.props.containers is from parent component in app.jsx. It is from a state of parent component. And I called function updateContainers to update parent's state. I think it works.

Reply:
This is still conceptually writing/changing a property of the `Containers` class. Note that it's *also* invalid to directly manipulate the state object (of yourself or your containing object),  this requires calling `setState()`. So you need to make a copy, with `newContainers = Object.assign({}, this.props.containers)`, or if you are into fancy new ES6, `newContainers = {...this.props.containers}` and call `setState()` on these.

However, this still has the problem that it makes assumptions on the effect of the called API function, and doesn't handle errors properly. It's much better to read the actual state from scratch after every API call that changes anything, until podman gets a proper notification API.

Reply:
Both containers and updateContainers() are props from parent component. I use updateContainers() to setState in parent component. Can child component use setSate to update parent's state from spread attributes {...} ? Could you give me some tutorial link to help me ?

Reply:
No, a component must never meddle with the state of any other component directly, only your own. If the `Containers` class triggers changes to the list of currently known containers, it must own the state. The valid way is to by and large do what you already do, and call a callback method (through a property like `this.props.updateContainers()`, but you must pass a new object (what I've shown with `Object.assign()` or the spread operator), not change your property object in-place.

But again, this is all too brittle. At the moment `podman.ListContainers()` and `GetContainerStats()` lives in app.jsx in `componentDidMount()`, i. e. just happens once. From then on your UI just tries to *predict* how the containers and states changed based on the actions from the UI. But these sometimes fail, containers can spontaneously stop or be started/stopped from other places than Cockpit, so the UI will quickly become out of date.

The proper way to fix that is to teach podman about a proper notification API. While it doesn't have one, I propose this architecture:

 * At the top level (app.jsx or maybe a container-state.es6 helper) you maintain the state of all containers, and a `updateContainers()` method (no arguments), which basically does what app.jsx' `componentDidMount()` does.

 * All components pass the app's `updateContainers()` down as a callback property to components.

 * Whenever any component calls a podman API function, it calls `this.props.updateContainers()` after it returns (on both success and failure!), so that the state and UI will update to reality.

Reply:
Oh, should be obviousy, but just in case: of course components would also pass down the actual container state object (or a relevant part of it)  as properties to its contained components. Then the right bits of the UI will update at the right time.